include/graaflib/algorithm/shortest_path.tpp

```diff
@@ -153,4 +153,45 @@ dijkstra_shortest_paths(const graph<V, E, T>& graph,
   return shortest_paths;
 }
 
+template <typename V, typename E, graph_type T, typename WEIGHT_T>
+std::unordered_map<vertex_id_t, graph_path<WEIGHT_T>>
+bellman_ford_shortest_paths(const graph<V, E, T>& graph, vertex_id_t start_vertex) {
+
+  std::unordered_map<vertex_id_t, graph_path<WEIGHT_T>> shortest_paths;
+
+  // Initialize the shortest path distances from the starting vertex to "infinity" for all vertices
+  for (const auto& vertex : graph.get_vertices()) {
+    shortest_paths[vertex.first] = {
+      {vertex.first},
+      std::numeric_limits<WEIGHT_T>::max()
+    };
+  }
+
+  // Set the distance from the starting vertex to itself to 0
+  shortest_paths[start_vertex] = {
+    {start_vertex}, 
+    0
+  };
+
+  // Relax edges for |V| - 1 iterations
+  for (std::size_t i = 1; i < graph.vertex_count(); ++i) {
+    for (const auto& edge : graph.get_edges()) {
+      const auto [u, v]{edge.first};
+      WEIGHT_T weight = get_weight(edge);
+
```

Comment:
Here, `graph.get_edges()` returns an `unordered_map` of `edge_id_t` to `edge_t`. Iterating over it means that the type of `edge` here is `std::pair<edge_id_t, edge_t>`. 

As a result, the call `get_weight(edge)` is not able to extract the actual weight of the edge, but instead will always return a unit weight. Instead, it should be called with the actual `edge_t:

```c++
WEIGHT_T weight = get_weight(edge.second);
```

To make this more clear in the code, I would also suggest to use structured binding:
```suggestion
    for (const auto& [edge_id, edge] : graph.get_edges()) {
      const auto [u, v]{edge_id};
      WEIGHT_T weight = get_weight(edge);

```

I assume this might also be the reason why the tests fail.